src/util/imagefiledata.cpp

```diff
@@ -0,0 +1,71 @@
+#include "util/imagefiledata.h"
+
+#include <QBuffer>
+#include <QByteArray>
+#include <QFile>
+#include <QImage>
+#include <QImageReader>
+
+#include "util/assert.h"
+
+ImageFileData::ImageFileData(
+        const QByteArray& coverArtBytes, const QByteArray& coverArtFormat)
+        : QImage(QImage::fromData(coverArtBytes)),
+          m_coverArtBytes(coverArtBytes) {
+    m_coverArtFormat = coverArtFormat.isNull() ? readFormatFrom(m_coverArtBytes) : coverArtFormat;
+}
+
+//static
+ImageFileData ImageFileData::fromFilePath(const QString& coverArtFilePath) {
+    QFile coverArtFile = QFile(coverArtFilePath);
+    coverArtFile.open(QIODevice::ReadOnly);
+    QByteArray coverArtBytes = coverArtFile.readAll();
+    coverArtFile.close();
+    QByteArray coverArtFormat = readFormatFrom(coverArtBytes);
+    return ImageFileData(coverArtBytes, coverArtFormat);
+}
+
+//static
+QByteArray ImageFileData::readFormatFrom(const QByteArray& coverArtBytes) {
+    QByteArray bytes = coverArtBytes;
+    QBuffer b = QBuffer(&bytes);
+    b.open(QIODevice::ReadOnly);
+    auto reader = QImageReader(&b);
+    QByteArray coverArtFormat = reader.format();
+    VERIFY_OR_DEBUG_ASSERT(!coverArtFormat.isEmpty()) {
+        return QByteArray{};
+    }
+    return coverArtFormat;
+}
+
+bool ImageFileData::saveFile(const QString& coverArtAbsoluteFilePath) const {
+    if (m_coverArtBytes.isNull()) {
+        return this->save(coverArtAbsoluteFilePath);
+    }
+
+    QFile coverArtFile(coverArtAbsoluteFilePath);
+    coverArtFile.open(QIODevice::WriteOnly);
+    if (!coverArtFile.write(m_coverArtBytes)) {
+        qWarning() << "Failed to write cover art"
+                   << coverArtAbsoluteFilePath;
+        return false;
+    }
+
+    coverArtFile.close();
+    return true;
+}
+
+bool ImageFileData::saveFileAddSuffix(const QString& pathWithoutSuffix) const {
+    VERIFY_OR_DEBUG_ASSERT(!m_coverArtBytes.isEmpty() || !m_coverArtFormat.isEmpty()) {
+        return false;
+    }
+    QFile coverArtFile(pathWithoutSuffix + '.' + m_coverArtFormat);
+    coverArtFile.open(QIODevice::WriteOnly);
```

Comment:
return value is not checked. We should do it to not hit assertions inside QT and crash Mixxx